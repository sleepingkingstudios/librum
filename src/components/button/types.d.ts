import type { IconDefinition } from '@fortawesome/fontawesome-common-types';

import {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  ButtonShims,
} from './shims';

import type { ColorTypes } from '../types';

type ButtonContents = {
  children: React.ReactNode,
} | {
  icon?: IconDefinition,
  label: string,
};

type ButtonOptions = {
  className?: string,
  children?: React.ReactNode,
  disabled?: boolean,
  icon?: IconDefinition,
  label?: string,
  outline?: boolean,
  size?: ButtonSize,
  type?: ColorTypes,
} & ButtonContents;

type ButtonSize = ("sm" | "md" | "lg");

type ButtonType = ("button" | "reset" | "submit");

export type LinkButtonProps = {
  url: string,
} & ButtonOptions;

export type PlainButtonProps = {
  htmlType?: ButtonType,
  onClick?: () => void,
} & ButtonOptions;

export type ButtonProps = LinkButtonProps | PlainButtonProps;
