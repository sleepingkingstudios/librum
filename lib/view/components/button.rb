# frozen_string_literal: true

module View::Components
  # Renders a button component.
  class Button < ViewComponent::Base
    # @param color [String] the color of the button.
    # @param label [String] the button label.
    # @param type [String] the type of button, either "button", "reset" or
    #   "submit".
    def initialize(color: nil, label: nil, type: 'button')
      super()

      @color = color
      @label = label
      @type  = type
    end

    # @return [String] the color of the button.
    attr_reader :color

    # @return [String] the button label.
    attr_reader :label

    # @return [String] the type of button, either "button", "reset" or "submit".
    attr_reader :type

    private

    def class_names
      names = %w[button]

      names << "is-#{color}" if color

      names.join(' ')
    end
  end
end
