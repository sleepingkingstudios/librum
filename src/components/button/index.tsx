import * as React from 'react';

import type { IconDefinition } from '@fortawesome/fontawesome-common-types';
import { FontAwesomeIcon as Icon } from '@fortawesome/react-fontawesome';
import { Link } from 'react-router-dom';

import { joinClassNames } from '@utils/react-utils';

import type {
  ButtonProps,
  LinkButtonProps,
  PlainButtonProps,
} from './types';

export type { ButtonProps } from './types';

const generateClassName = ({
  className,
  outline = false,
  size,
  type,
}: {
  className?: string,
  outline: boolean,
  size?: string,
  type?: string,
}): string => joinClassNames(
  'btn',
  outline ? 'btn-outline' : null,
  size? `btn-${size}` : null,
  type ? `btn-${type}` : null,
  className,
);

const renderContents = ({
  children,
  icon,
  label,
}: {
  children?: React.ReactNode,
  icon?: IconDefinition,
  label?: string,
}): React.ReactNode => {
  if (children) { return children; }

  if (label) {
    if (!icon) { return label; }

    return (
      <>
        <Icon className="pr-2" icon={icon} />

        { label }
      </>
    )
  }

  return null;
};

const LinkButton = ({
  children,
  className,
  label,
  icon,
  outline,
  size,
  type,
  url,
}: LinkButtonProps): JSX.Element => {
  const joinedClassName = generateClassName({
    className,
    outline,
    size,
    type,
  });

  return (
    <Link className={joinedClassName} to={url}>
      { renderContents({ children, icon, label }) }
    </Link>
  );
};

const PlainButton = ({
  children,
  className,
  disabled = false,
  htmlType = 'button',
  label,
  icon,
  onClick,
  outline,
  size,
  type,
}: PlainButtonProps): JSX.Element => {
  const joinedClassName = generateClassName({
    className,
    outline,
    size,
    type,
  });

  return (
    <button
      className={joinedClassName}
      disabled={disabled}
      onClick={onClick}
      type={htmlType}
    >
      { renderContents({ children, icon, label }) }
    </button>
  );
};

export const Button = (props: ButtonProps): JSX.Element => {
  const { disabled } = props;

  if (('url' in props) && !disabled) { return (<LinkButton {...props} />); }

  return (<PlainButton {...props} />);
};
