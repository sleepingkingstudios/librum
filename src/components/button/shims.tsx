import * as React from 'react';

export const ButtonShims = () => {
  <>
    <button className="btn btn-outline" />
    <button className="btn btn-primary.btn-outline" />
    <button className="btn btn-danger.btn-outline" />
    <button className="btn btn-info.btn-outline" />
    <button className="btn btn-muted.btn-outline" />
    <button className="btn btn-success.btn-outline" />
    <button className="btn btn-warning.btn-outline" />
    <button className="btn btn-sm" />
    <button className="btn btn-md" />
    <button className="btn btn-lg" />
  </>
};
