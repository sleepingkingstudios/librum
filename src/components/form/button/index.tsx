import * as React from 'react';

import { joinClassNames } from '@utils/react-utils';
import type { UpToTwelveColumns } from '@components/types';

interface IFormButtonProps {
  className?: string;
  children: React.ReactNode;
  disabled?: boolean;
  cols?: UpToTwelveColumns;
  type?: "button" | "reset" | "submit";
}

const columnSpanClassName = ({ cols }: { cols: number }): string | null => {
  if (cols === null) { return null; }

  return `col-span-${cols}`;
};

export const FormButton = ({
  className,
  children,
  disabled = false,
  cols = null,
  type = 'button',
}: IFormButtonProps): JSX.Element => {
  const joinedClassName = joinClassNames(
    'form-button',
    columnSpanClassName({ cols }),
    className,
  );

  return (
    <button
      disabled={disabled}
      className={joinedClassName}
      type={type}
    >
      { children }
    </button>
  );
};
