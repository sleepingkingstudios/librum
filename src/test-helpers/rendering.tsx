import * as React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import {
  MemoryRouter,
  Routes,
  Route
} from "react-router-dom";
import {
  RenderOptions,
  RenderResult,
  render,
} from '@testing-library/react';

import {
  createStore,
  storeType,
} from './store';
import {
  ThemeContext,
  defaultTheme,
} from '@themes/context';
import type { Theme } from '@themes';

interface customRenderOptions extends RenderOptions {
  initialEntries?: Array<string>;
  router?: true | withRouterOptions;
  store?: (storeType | true);
  theme?: (Theme | true);
}

interface withRouterOptions {
  initialEntries?: Array<string>;
  path?: string;
}

export const withRouter = (
  component: React.ReactElement,
  {
    initialEntries = ['/'],
    path,
  }: withRouterOptions,
): React.ReactElement => {
  let wrapped = component;

  if (path) {
    wrapped = (
      <Routes>
        <Route path={path} element={wrapped} />
      </Routes>
    );
  }

  return (
    <MemoryRouter initialEntries={initialEntries}>
      { wrapped }
    </MemoryRouter>
  );
};

export const withStore = (
  component: React.ReactElement,
  { store }: { store: (storeType | true) }
): React.ReactElement => {
  let configuredStore: storeType;

  if (store === true) {
    configuredStore = createStore().store;
  } else {
    configuredStore = store;
  }

  return (
    <ReduxProvider store={configuredStore}>
      { component }
    </ReduxProvider>
  );
};

export const withTheme = (
  component: React.ReactElement,
  { theme }: { theme: (Theme | true) }
): React.ReactElement => {
  let configuredTheme: Theme;

  if (theme === true) {
    configuredTheme = defaultTheme;
  } else {
    configuredTheme = theme;
  }

  return (
    <ThemeContext.Provider value={configuredTheme}>
      { component }
    </ThemeContext.Provider>
  );
};

const customRender = (
  component: React.ReactElement,
  options?: customRenderOptions,
  strictMode = true,
): RenderResult => {
  let wrapped: React.ReactElement = component;

  if (strictMode) {
    wrapped = (<React.StrictMode>{ wrapped }</React.StrictMode>);
  }

  if (!options) { return render(wrapped); }

  if ('router' in options && options.router) {
    if (options.router === true) {
      wrapped = withRouter(wrapped, { initialEntries: options.initialEntries });
    } else {
      wrapped = withRouter(
        wrapped,
        {
          initialEntries: options.initialEntries,
          ...options.router,
        },
      );
    }
  }

  if ('store' in options) {
    wrapped = withStore(wrapped, { store: options.store });
  }

  if ('theme' in options) {
    wrapped = withTheme(wrapped, { theme: options.theme });
  }

  return render(wrapped, options);
};

export { customRender as render };
