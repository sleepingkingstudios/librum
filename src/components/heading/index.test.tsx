import * as React from 'react';

import '@testing-library/jest-dom';

import { Heading } from './index';
import { ButtonProps as Button } from '@components/button';
import { render } from '@test-helpers/rendering';

describe('<Heading />', () => {
  it('should render the contents', () => {
    const { getByRole } = render(
      <Heading size={1}>Greetings, Programs</Heading>,
    );

    const heading = getByRole('heading');

    expect(heading).toHaveTextContent('Greetings, Programs');
  });

  it('should match the snapshot', () => {
    const { asFragment } = render(
      <Heading size={1}>Greetings, Programs</Heading>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe('with buttons: value', () => {
    const buttons: Button[] = [
      {
        label: 'Go',
        type: 'primary',
      },
      {
        label: 'Stop',
        type: 'danger',
      },
    ];

    it('should render the buttons', () => {
      const { getAllByRole } = render(
        <Heading buttons={buttons} size={1}>Greetings, Programs</Heading>,
      );
      const expected = ['Go', 'Stop'];

      const rendered = getAllByRole('button');

      expect(rendered).toHaveLength(2);
      expect(rendered.map(btn => btn.textContent)).toEqual(expected);
    });

    it('should render the contents', () => {
      const { getByRole } = render(
        <Heading buttons={buttons} size={1}>Greetings, Programs</Heading>,
      );

      const heading = getByRole('heading');

      expect(heading).toHaveTextContent('Greetings, Programs');
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <Heading buttons={buttons} size={1}>Greetings, Programs</Heading>,
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with size: 2', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(
        <Heading size={2}>Greetings, Programs</Heading>,
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with size: 3', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(
        <Heading size={3}>Greetings, Programs</Heading>,
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with size: 4', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(
        <Heading size={4}>Greetings, Programs</Heading>,
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with size: 5', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(
        <Heading size={5}>Greetings, Programs</Heading>,
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with size: 6', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(
        <Heading size={6}>Greetings, Programs</Heading>,
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
