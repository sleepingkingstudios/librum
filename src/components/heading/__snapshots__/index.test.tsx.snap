// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Heading /> should match the snapshot 1`] = `
<DocumentFragment>
  <h1>
    Greetings, Programs
  </h1>
</DocumentFragment>
`;

exports[`<Heading /> with buttons: value should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col sm:flex-row"
  >
    <h1>
      Greetings, Programs
    </h1>
    <ul
      class="justify-end flex flex-col grow sm:flex-row btn-list"
    >
      <button
        class="btn btn-primary"
        type="button"
      >
        Go
      </button>
      <button
        class="btn btn-danger"
        type="button"
      >
        Stop
      </button>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`<Heading /> with size: 2 should match the snapshot 1`] = `
<DocumentFragment>
  <h2>
    Greetings, Programs
  </h2>
</DocumentFragment>
`;

exports[`<Heading /> with size: 3 should match the snapshot 1`] = `
<DocumentFragment>
  <h3>
    Greetings, Programs
  </h3>
</DocumentFragment>
`;

exports[`<Heading /> with size: 4 should match the snapshot 1`] = `
<DocumentFragment>
  <h4>
    Greetings, Programs
  </h4>
</DocumentFragment>
`;

exports[`<Heading /> with size: 5 should match the snapshot 1`] = `
<DocumentFragment>
  <h5>
    Greetings, Programs
  </h5>
</DocumentFragment>
`;

exports[`<Heading /> with size: 6 should match the snapshot 1`] = `
<DocumentFragment>
  <h6>
    Greetings, Programs
  </h6>
</DocumentFragment>
`;
